### PR TITLE
Allow changing settings at the client level.

### PR DESCRIPTION
**Pull request by rozza**
Added MongoClientOperations to handle client level operations and to allow for changable settings.

JAVA-5259

### CODE REVIEW

**Comment by rozza** [comment]
I ended up choosing `MongoClientOperations` as a name but happy to change if theres a better alternative.

**Review by jyemin — CHANGES_REQUESTED** [review]
All the changes look good.  I just have a few comments

1. We should have some tests of the new with* methods.  Similar to `com.mongodb.client.internal.MongoDatabaseSpecification#should behave correctly when using withWriteConcern`, as one example.
2. I don't love the name `MongoClientOperations`.  This would be the first use of the word `Operations` in the API, and I'd rather not introduce new terminology in this one place. so I'll throw my favorites in the hat.  To be more consistent with `MongoDatabase` and `MongoCollection`,  `MongoDeployment` or `MongoCluster` would be more apt.  The first is more consistent with our general documentation, e.g. https://www.mongodb.com/docs/atlas/create-database-deployment/, while the second is more consistent with existing driver API, e.g. `ClusterDescription`, and also doesn't introduce a new word (Deployment) into the API.

File: driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedCrudHelper.java [hunk 1]
```diff
@@ -38,7 +38,7 @@
 import com.mongodb.client.ListIndexesIterable;
 import com.mongodb.client.ListSearchIndexesIterable;
 import com.mongodb.client.MongoChangeStreamCursor;
-import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoCluster;
```
Comment (jyemin) [inline driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedCrudHelper.java:41]: Changes to this file LGTM, but are there skipped tests that can now be run as a result of the changes?

Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/unified/UnifiedCrudHelper.java:41]: Yes, the removal of: 
`assumeFalse("TODO (CSOT) - JAVA-5259 No client.withTimeout", testDescription.endsWith("on client"));` in ClientSideOperationTimeoutTest, means that tests are now run using this path.

**Review by vbabanin — CHANGES_REQUESTED** [review]

File: driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/MongoCluster.java [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+ * </p>
+ *
+ * @see MongoClient
+ * @since CSOT
+ */
+@Immutable
+public interface MongoCluster {
+
+    /**
+     * Get the codec registry for the MongoClientClusterOperations.
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/MongoCluster.java:49]: Done - copypasta :(

File: driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/Crypts.java [hunk 3]
```diff
@@ -41,28 +40,28 @@ public final class Crypts {
     private Crypts() {
     }
 
-    public static Crypt createCrypt(final MongoClientImpl client, final AutoEncryptionSettings settings) {
+    public static Crypt createCrypt(final MongoClientSettings mongoClientSettings, final AutoEncryptionSettings autoEncryptionSettings) {
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/Crypts.java:43]: Done

File: driver-kotlin-coroutine/src/main/kotlin/com/mongodb/kotlin/client/coroutine/MongoCluster.kt [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+     *     - `maxCommitTimeMS`: The maximum amount of time to allow a single `commitTransaction` command to execute.
+     * - `0` means infinite timeout.
+     * - `> 0` The time limit to use for the full execution of an operation.
+     *
+     * @return the optional timeout duration
+     */
+    public fun timeout(timeUnit: TimeUnit = TimeUnit.MILLISECONDS): Long? = wrapped.getTimeout(timeUnit)
+
+    /**
+     * Create a new MongoDatabase instance with a different codec registry.
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-kotlin-coroutine/src/main/kotlin/com/mongodb/kotlin/client/coroutine/MongoCluster.kt:79]: Done

File: driver-sync/src/test/functional/com/mongodb/client/ClientSideOperationTimeoutTest.java [hunk 5]
```diff
@@ -82,8 +82,6 @@ public static void checkSkipCSOTTest(final String fileDescription, final String
         assumeFalse("No count command helper", testDescription.endsWith("count on collection"));
         assumeFalse("No operation based overrides", fileDescription.equals("timeoutMS can be overridden for an operation"));
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-sync/src/test/functional/com/mongodb/client/ClientSideOperationTimeoutTest.java:84]: I think we can still ignore this as its for testing per operation based timeouts to ensure they override the client/database/collection settings. We don't have an API for doing that - eg there is no `listDatabaseNames(options)` that takes a timeout override.

As there are already tests at the client, database and collection level where we test operations with the expected timeoutMS.

**Comment by rozza** [comment]
@vbabanin @jyemin this is ready for re-review. Test failures are related to `rangePreview` and others are flaky / racey tests that are scheduled to be reviewed as part of JAVA-5104.

**Review by jyemin — APPROVED** [review]
LGTM

**Review by vbabanin — APPROVED** [review]

**Comment by vbabanin** [comment]
LGTM!

**Pull request merged**